File: src/dotnet/UI.Blazor/Components/VirtualList/VirtualListDataQuery.cs
namespace ActualChat.UI.Blazor.Components;

public sealed class VirtualListDataQuery(Range<string> keyRange, Range<double> virtualRange, Range<int> moveRange)
{
    public static readonly VirtualListDataQuery None = new(default ,default, default);

    public Range<string> KeyRange { get; } = keyRange;
    public Range<double> VirtualRange { get; } = virtualRange;
    public Range<int> MoveRange { get; } = moveRange;

    public int? ExpectedCount { get; init; }

    public bool IsNone
        => ReferenceEquals(this, None);

    public override string ToString()
        => $"⁇({KeyRange} -> ~{MoveRange})";
}

File: src/dotnet/UI.Blazor/Components/VirtualList/ts/virtual-list-data-query.ts
import { NumberRange, Range } from './range';

export class VirtualListDataQuery
{
    public static None: VirtualListDataQuery = new VirtualListDataQuery(
        new Range<string>('', ''),
        new NumberRange(0, 0),
        new NumberRange(0, 0));

    public expectedCount?: number;

    constructor(public keyRange: Range<string>, public virtualRange: NumberRange, public moveRange: NumberRange)
    { }

    public get isNone(): boolean {
        return this === VirtualListDataQuery.None;
    }
}
